gym-unity/gym_unity/envs/__init__.py

```diff
@@ -368,52 +372,69 @@ def _sanitize_info(self, step_result: BatchedStepResult) -> BatchedStepResult:
                 "The number of agents in the scene does not match the expected number."
             )
 
-        # remove the done Agents
-        indices_to_keep: List[int] = []
-        for index, is_done in enumerate(step_result.done):
-            if not is_done:
-                indices_to_keep.append(index)
+        if step_result.n_agents() - sum(step_result.done) != self._n_agents:
+            raise UnityGymException(
+                "The number of agents in the scene does not match the expected number."
+            )
+
+        for index, agent_id in enumerate(step_result.agent_id):
+            if step_result.done[index]:
+                gym_index = self._gym_id_order.index(agent_id)
+                self._done_agents_index_to_last_reward[gym_index] = step_result.reward[
+                    index
+                ]
+                self._gym_id_order[gym_index] = -1  # no agent at that index
 
         # Set the new AgentDone flags to True
         # Note that the corresponding agent_id that gets marked done will be different
         # than the original agent that was done, but this is OK since the gym interface
         # only cares about the ordering.
         for index, agent_id in enumerate(step_result.agent_id):
             if not self._previous_step_result.contains_agent(agent_id):
+                # insert the id in the current id list:
+                original_index = self._gym_id_order.index(-1)
+                self._gym_id_order[original_index] = agent_id
+                # This is a new agent
                 step_result.done[index] = True
-            if agent_id in self._done_agents:
-                step_result.done[index] = True
-        self._done_agents = set()
+                # The index of the agent among not-done agents is
+                original_index = self._gym_id_order.index(agent_id)
```

Comment:
Sorry, should have been more explicit in the previous comment. `original_index` here should be the same as the one above (since you just set `self._gym_id_order[original_index] = agent_id`), so you don't need to compute it here too